packages/client/src/internals/executeChain.ts

```diff
@@ -0,0 +1,59 @@
+import { observable } from './observable';
+import {
+  OperationLink,
+  Operation,
+  OperationResult,
+  PrevCallback,
+} from '../links/core';
+
+export function executeChain(opts: { links: OperationLink[]; op: Operation }) {
+  const $result = observable<OperationResult | null>(null);
+  const $destroyed = observable(false);
+
+  function walk({
+    index,
+    op,
+    stack,
+  }: {
+    index: number;
+    op: Operation;
+    stack: PrevCallback[];
+  }) {
+    const link = opts.links[index];
+    const prev: PrevCallback =
+      index === 0 ? (value) => $result.set(value) : stack[index - 1];
+
+    link({
+      op,
+      prev,
+      next: (op, prevOp) => {
+        const prevStack = stack.slice();
+        prevStack[index] = prevOp;
+        walk({ index: index + 1, op, stack: prevStack });
+      },
+      onDestroy: (callback) => {
+        const unsub = $destroyed.subscribe((aborted) => {
+          if (aborted) {
+            callback();
+            unsub();
+          }
+        });
+      },
+    });
+  }
+  walk({ index: 0, op: opts.op, stack: [] });
```

Comment:
What happens if you go prev more than once, is the observable `$result` correct?

Reply:
yes, here's a test: 423524d

the thing that _would_ cause problems is links that would be calling down multiple times but could also be re-instantiated by something calling `next()` several times.. so the order is a bit fragile, but that just feels like the nature of the beast -- i.e. you can't put a `cacheLink` (that could call `prev()` several times on cache changes) after a `retryLink` that'd re-instantiate it on failures.

not sure if this is an issue or not yet.

Reply:
after struggling with this a while and trying to figure out a solution i realised it could might as well be added on the ending link where the data is loaded whenever this is needed